src/Microsoft.ApiDesignGuidelines.Analyzers/Core/UsePropertiesWhereAppropriate.cs

```diff
@@ -1,39 +1,94 @@
+                                                                         s_localizableMessage,
+                                                                         DiagnosticCategory.Design,
+                                                                         DiagnosticSeverity.Warning,
+                                                                         isEnabledByDefault: false,
+                                                                         description: s_localizableDescription,
+                                                                         helpLinkUri: "http://msdn.microsoft.com/library/ms182181.aspx",
+                                                                         customTags: WellKnownDiagnosticTags.Telemetry);
+        private const string GetHashCodeName = "GetHashCode";
+        private const string GetEnumeratorName = "GetEnumerator";
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
-        { 
-            
+        {
+            analysisContext.RegisterOperationBlockStartAction(context =>
+            {
+                var methodSymbol = context.OwningSymbol as IMethodSymbol;
+
+                if (methodSymbol == null ||
+                    methodSymbol.ReturnsVoid ||
+                    methodSymbol.ReturnType.Kind == SymbolKind.ArrayType ||
+                    methodSymbol.Parameters.Length > 0 ||
+                    !(methodSymbol.DeclaredAccessibility == Accessibility.Public || methodSymbol.DeclaredAccessibility == Accessibility.Protected) ||
+                    methodSymbol.IsAccessorMethod() ||
+                    !IsPropertyLikeName(methodSymbol.Name))
+                {
+                    return;
+                }
+
+                // A few additional checks to reduce the noise for this diagnostic:
+                // Ensure that the method is non-generic, non-virtual/override, has no overloads and doesn't have special names: 'GetHashCode' or 'GetEnumerator'.
+                // Also avoid generating this diagnostic if the method body has any invocation expressions.
+                if (methodSymbol.IsGenericMethod ||
+                    methodSymbol.IsVirtual ||
+                    methodSymbol.IsOverride ||
+                    methodSymbol.ContainingType.GetMembers(methodSymbol.Name).Length > 1 ||
+                    methodSymbol.Name == GetHashCodeName ||
+                    methodSymbol.Name == GetEnumeratorName)
+                {
+                    return;
+                }
+                
+                bool hasInvocations = false;
+                context.RegisterOperationAction(operationContext => 
+                {
+                    hasInvocations = true;
+                }, OperationKind.InvocationExpression);
```

Comment:
Do we care about unary/binary/conversion operations that call operation methods? How about object creations that call constructors?